hw-model/src/openocd/openocd_server.rs

```diff
@@ -0,0 +1,228 @@
+            });
+            std::thread::spawn(move || {
+                printer::accumulate(stderr, "OpenOCD::stderr", Default::default())
+            });
+        }
+
+        let kill_guard = scopeguard::guard(child, |mut child| {
+            let _ = child.kill();
+        });
+
+        if log_stdio {
+            println!("Connecting to OpenOCD Tcl interface ...");
+        }
+        let stream = TcpStream::connect(("localhost", openocd_port))
+            .context("failed to connect to OpenOCD socket")?;
+        let connection = Self {
+            server_process: scopeguard::ScopeGuard::into_inner(kill_guard),
+            reader: BufReader::new(stream.try_clone()?),
+            writer: stream,
+        };
+
+        Ok(connection)
+    }
+
+    /// Send a string to the OpenOCD Tcl server.
+    fn send(&mut self, cmd: &str) -> Result<()> {
+        // The protocol is to send the command followed by a `0x1a` byte,
+        // see https://openocd.org/doc/html/Tcl-Scripting-API.html#Tcl-RPC-server
+
+        // Sanity check to ensure that the command string is not malformed.
+        if cmd.contains('\x1A') {
+            bail!("Tcl command string should be contained inside the string to send.");
+        }
+        self.writer
+            .write_all(cmd.as_bytes())
+            .context("failed to send a command to the OpenOCD server")?;
+        self.writer
+            .write_all(&[0x1a])
+            .context("failed to send the command terminator to OpenOCD server")?;
+        self.writer.flush().context("failed to flush stream")?;
+        Ok(())
+    }
+
+    /// Receive a string from the OpenOCD Tcl server.
+    fn recv(&mut self) -> Result<String> {
+        let mut buf = Vec::new();
+        self.reader.read_until(0x1A, &mut buf)?;
+        if !buf.ends_with(b"\x1A") {
+            bail!(OpenOcdError::PrematureExit);
+        }
+        buf.pop();
+        String::from_utf8(buf).context("failed to parse OpenOCD response as UTF-8")
+    }
+
+    /// Execute a Tcl command via the OpenOCD and wait for its response.
+    pub fn execute(&mut self, cmd: &str) -> Result<String> {
+        self.send(cmd)?;
+        self.recv()
+    }
+
+    pub fn shutdown(mut self) -> Result<()> {
+        self.execute("shutdown")?;
+        self.server_process
+            .wait()
+            .context("failed to wait for OpenOCD server to exit")?;
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+
+    #[test]
+    fn test_openocd_server() {
+        if let Ok(mut openocd) =
+            OpenOcdServer::spawn(Path::new("openocd"), /*log_stdio=*/ true)
+        {
+            if let Ok(version) = openocd.execute("version") {
+                println!("OpenOCD version: {version}");
+            } else {
+                panic!("Failed to execute an openocd command.");
+            }
+            assert!(openocd.shutdown().is_ok());
+        } else {
+            panic!("Failed to spawn an openocd subprocess.");
+        }
```

Comment:
nit: this might be more readable using let else syntax:
```suggestion
        let Ok(mut openocd) =
            OpenOcdServer::spawn(Path::new("openocd"), /*log_stdio=*/ true) else {
            panic!("Failed to spawn an openocd subprocess.");
        }
        let Ok(version) = openocd.execute("version") else {
            panic!("Failed to execute an openocd command.");
        }
        println!("OpenOCD version: {version}");
        assert!(openocd.shutdown().is_ok());
```